src/validations/__tests__/runRules.test.js

```diff
@@ -0,0 +1,177 @@
+import runRules from '../html/runRules';
+
+describe('structural html validation', () => {
+  test('openTag row', async () => {
+    expect.assertions(1);
+    await runRules(
+      [
+        {
+          openTag({row}) {
+            expect(row).toEqual(2);
+          },
+          *done() {},
+        },
+      ],
+      '\n\n<div>',
+    );
+  });
+
+  test('openTag column', async () => {
+    expect.assertions(1);
+    await runRules(
+      [
+        {
+          openTag({column}) {
+            expect(column).toEqual(1);
+          },
+          *done() {},
+        },
+      ],
+      '\n <div>',
+    );
+  });
+  test('openTag name', async () => {
+    expect.assertions(1);
+    await runRules(
+      [
+        {
+          openTag(_, {name}) {
+            expect(name).toEqual('div');
+          },
+          *done() {},
+        },
+      ],
+      '<div>',
+    );
+  });
+
+  test('openTag skips void tag', () => {
+    expect(async () => {
+      await runRules(
+        [
+          {
+            openTag(location, tag) {
+              throw new Error(
+                `location: ${JSON.stringify(location)} tag: ${JSON.stringify(
+                  tag,
+                )}`,
+              );
+            },
+            *done() {},
+          },
+        ],
+        '<img>',
+      );
+    }).not.toThrow();
+  });
+
+  test('openTag skips self-closing non-void tag', async () => {
+    expect(async () => {
```

Comment:
Similar situation here, I don’t think `expect(…).not.toThrow()` would work out of the box when passed an async function—it looks like you can use [`.rejects`](https://jestjs.io/docs/en/expect#rejects), and return/await the result of the `expect`, to do what you want here.

Reply:
Had to read the original runRules code for this. As long as the `await runRules` finishes successfully, the test is valid because it's testing that a tag was skipped.